### PR TITLE
Addition and multiplication over cuarray and cusparse

### LINKED ISSUE #1113: Addition and multiplication over cuarray and cusparse generates inconsistent outcome types or error
```julia
julia> using CUDA, SparseArrays

julia> dense = rand([0,1], 3, 4)  # Matrix{Int64}

julia> sp = sparse(dense)  # SparseMatrixCSC{Int64, Int64}

julia> cuarray = cu(dense)  # CuArray{Int64, 2, CUDA.Mem.DeviceBuffer}

julia> cusparse = cu(sp);  # CUDA.CUSPARSE.CuSparseMatrixCSC{Int64}
```

It is reasonable to add dense array and sparse array and gives a dense array.

```julia
julia> typeof(dense + sp)
Matrix{Int64} (alias for Array{Int64, 2})

julia> typeof(sp + dense)
Matrix{Int64} (alias for Array{Int64, 2})
```

But, in cuda, adding cuarray and cusparse array gives a error related to broadcast.

```julia
julia> typeof(cuarray + cusparse)
ERROR: GPU compilation of kernel broadcast_kernel(CUDA.CuKern...

julia> typeof(cusparse + cuarray)
ERROR: GPU compilation of kernel broadcast_kernel(CUDA.CuKern...
```

Adding two cuda sparse arrays together gives a undefined error.

```julia
julia> typeof(cusparse + cusparse)
ERROR: UndefVarError: geam not defined
Stacktrace:
 [1] +(A::CUDA.CUSPARSE.CuSparseMatrixCSC{Int64}, B::CUDA.CUSPARSE.CuSparseMatrixCSC{Int64})
   @ CUDA.CUSPARSE ~/.julia/packages/CUDA/VGl9W/lib/cusparse/interfaces.jl:83
```

If we check matrix multiplication, it gives different types of array.

```julia
julia> cusparse = cu(sp')
4×3 adjoint(::CUDA.CUSPARSE.CuSparseMatrixCSC{Int64}) with eltype Int64:
┌ Warning: Performing scalar indexing on task Task (runnable) @0x00007f0e555cc160.
│ Invocation of getindex resulted in scalar indexing of a GPU array.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on the GPU, but very slowly on the CPU,
│ and therefore are only permitted from the REPL for prototyping purposes.
│ If you did intend to index this array, annotate the caller with @allowscalar.
└ @ GPUArrays ~/.julia/packages/GPUArrays/UBzTm/src/host/indexing.jl:56
 1  1  1
 0  1  0
 0  0  1
 1  1  0

julia> typeof(cuarray * cusparse)
┌ Warning: Performing scalar indexing on task Task (runnable) @0x00007f0e54d8c010.
│ Invocation of getindex resulted in scalar indexing of a GPU array.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on the GPU, but very slowly on the CPU,
│ and therefore are only permitted from the REPL for prototyping purposes.
│ If you did intend to index this array, annotate the caller with @allowscalar.
└ @ GPUArrays ~/.julia/packages/GPUArrays/UBzTm/src/host/indexing.jl:56
Matrix{Int64} (alias for Array{Int64, 2})

julia> typeof(cusparse * cuarray)
CuArray{Int64, 2, CUDA.Mem.DeviceBuffer}

julia> typeof(cusparse * cu(sp))
Matrix{Int64} (alias for Array{Int64, 2})
```

This issue is partially reported in JuliaGPU/CUDA.jl#829.

System information:

```
julia> versioninfo()
Julia Version 1.6.2
Commit 1b93d53fc4 (2021-07-14 15:36 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 12
```

CUDA.jl version information:

```
(@v1.6) pkg> st CUDA
      Status `~/.julia/environments/v1.6/Project.toml`
  [052768ef] CUDA v3.4.1
```

### PR DESCRIPTION
**Pull request by yuehhua**
Closes #1113

### CODE REVIEW

**Comment by yuehhua** [comment]
@maleadt Is the CI working?

**Comment by maleadt** [comment]
Yes? Why wouldn't it be?

compute-sanitizer seems to fine an issue with some conversion:

```
[37mcusparse/interfaces[39m[37m                   (2) |        started at 2021-08-29T08:49:39.240[39m

      From worker 2:	========= Invalid __global__ write of size 4 bytes

      From worker 2:	=========     at 0x570 in void convert_CooToCsr_kernel<int=1>(int const *,int,int,int*)

      From worker 2:	=========     by thread (3,0,0) in block (0,0,0)

      From worker 2:	=========     Address 0x338f75c18 is out of bounds

      From worker 2:	========= 

      From worker 2:	========= Invalid __global__ write of size 4 bytes

      From worker 2:	=========     at 0x570 in void convert_CooToCsr_kernel<int=1>(int const *,int,int,int*)

      From worker 2:	=========     by thread (4,0,0) in block (0,0,0)

      From worker 2:	=========     Address 0x338f75c1c is out of bounds

      From worker 2:	========= 
```

**Comment by maleadt** [comment]
I take it this also fixes https://github.com/JuliaGPU/CUDA.jl/issues/528?

**Comment by yuehhua** [comment]
Sure! It fixes #528.

**Comment by yuehhua** [comment]
It seems to fail on v1.6 (debug)?

**Comment by maleadt** [comment]
> It seems to fail on v1.6 (debug)?

Right, so there's something wrong with this PR or some of the kernels it calls. See the error log I posted.

**Comment by yuehhua** [comment]
It seems like error occurs while freeing buffer?

```
      From worker 2:	 [14] unsafe_free!(xs::CuArray{UInt8, 1, CUDA.Mem.DeviceBuffer}, stream::CuStream)
      From worker 2:	    @ CUDA /media/yuehhua/Workbench/workspace/CUDA.jl/src/array.jl:80
      From worker 2:	 [15] unsafe_free!(xs::CuArray{UInt8, 1, CUDA.Mem.DeviceBuffer})
      From worker 2:	    @ CUDA /media/yuehhua/Workbench/workspace/CUDA.jl/src/array.jl:72
      From worker 2:	 [16] with_workspace(f::CUDA.CUSPARSE.var"#1053#1055"{Float32, CuSparseMatrixCSR{Float32}, Float32, CuSparseMatrixCSR{Float32}, CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, CUDA.CUSPARSE.CuMatrixDescriptor, CUDA.CUSPARSE.CuMatrixDescriptor, CUDA.CUSPARSE.CuMatrixDescriptor, Int64, Int64}, eltyp::Type{UInt8}, size::CUDA.CUSPARSE.var"#bufferSize#1054"{Float32, CuSparseMatrixCSR{Float32}, Float32, CuSparseMatrixCSR{Float32}, CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, CUDA.CUSPARSE.CuMatrixDescriptor, CUDA.CUSPARSE.CuMatrixDescriptor, CUDA.CUSPARSE.CuMatrixDescriptor, Int64, Int64}, fallback::Nothing)
      From worker 2:	    @ CUDA.APIUtils /media/yuehhua/Workbench/workspace/CUDA.jl/lib/utils/call.jl:79
      From worker 2:	 [17] with_workspace (repeats 2 times)
      From worker 2:	    @ /media/yuehhua/Workbench/workspace/CUDA.jl/lib/utils/call.jl:53 [inlined]
      From worker 2:	 [18] geam(alpha::Float32, A::CuSparseMatrixCSR{Float32}, beta::Float32, B::CuSparseMatrixCSR{Float32}, index::Char)
      From worker 2:	    @ CUDA.CUSPARSE /media/yuehhua/Workbench/workspace/CUDA.jl/lib/cusparse/extra.jl:35
```

**Comment by maleadt** [comment]
No, that's unrelated. The error code only triggers there because it's the first synchronous call. The real issue is spotted by compute-sanitizer, and is the error I linked above.

**Comment by yuehhua** [comment]
I still cannot find where `convert_CooToCsr_kernel` is called.

**Comment by maleadt** [comment]
It's the name of a kernel, probably an internal one in CUSPARSE. We don't currently have host backtraces (disabled due to a compute-santizer bug), so can't report the Julia function that called the kernel.

Anyway, this is why I was looking into upgrading the CUDA toolkit used for these debug runs in https://github.com/JuliaGPU/CUDA.jl/pull/950, since maybe the issue is with CUDA (i.e. not us wrongly calling that kernel).

**Comment by yuehhua** [comment]
COO + CSR or CSR + COO will trigger the bug, so I temporally turn down the tests.

File: lib/cusparse/interfaces.jl [hunk 1]
```diff
@@ -79,10 +79,87 @@ for (taga, untaga) in tag_wrappers, (wrapa, transa, unwrapa) in op_wrappers
...
+function Base.:(-)(A::Transpose{T,<:CuSparseMatrixCSR}, B::Transpose{T,<:CuSparseMatrixCSR}) where {T}
+    C = geam(one(T), A.parent, -one(T), B.parent, 'O')
+    cscC = CuSparseMatrixCSC(C)
+    return CuSparseMatrixCSR(cscC.colPtr, cscC.rowVal, cscC.nzVal, size(cscC))
+end
+
+function Base.:(+)(A::CuSparseMatrixCSR, B::CuSparseMatrix)
+    csrB = CuSparseMatrixCSR(B)
+    return geam(one(eltype(A)), A, one(eltype(A)), csrB, 'O')
+end
```
Comment (yuehhua) [inline lib/cusparse/interfaces.jl:142]: Maybe the `CuSparseMatrixCSR(B)` constructor and when `B` is a `CuSparseMatrixCOO` will trigger the bug.

Comment (yuehhua) [reply lib/cusparse/interfaces.jl:142]: It seems like occur in `cusparseXcoo2csr`.

**Comment by yuehhua** [comment]
It's ready for review.

**Comment by maleadt** [comment]
Rebased on top of the PR with CI fixes.

**Pull request closed without merging**